### ISSUE TITLE
Need a way to access Tuple elements and slice

### ISSUE THREAD

**Issue by ggilley** [fdb-rs/fdb#30]
I'm trying to read a block of tuples in and group them by a prefix. I need to be able to, for example, drop the first element of the tuple, then pass it on. It looks like you tried to prevent that unlike the other fdb apis? (e.g. java and golang).

I'm doing something like what is described here: https://forums.foundationdb.org/t/best-practice-of-storing-structs-should-i-pack-or-store-fields-separately/425/2?u=dongxineric

I'm storing the fields separately which means I need to be able to gather them by the "key" prefix.

**Comment by rajivr** [comment]
> I'm trying to read a block of tuples in and group them by a prefix. I need to be able to, for example, drop the first element of the tuple, then pass it on. It looks like you tried to prevent that unlike the other fdb apis? (e.g. java and golang).

Thanks for opening this issue. 

If I understand your issue correctly, what you are asking for is to take a tuple value of the form `(("name", "john"), ("age", 34),)` and drop the first element to create a tuple of the form `(("age", 34),)`.

If that is so, then currently there is no public API that can address it. I think this feature will be useful for layer development and I'll add it in the next release. 

I'll outline my initial thoughts on how we could implement this in case you are interested in contributing a PR. :-)

Internally, Tuple elements are stored as a `Vec<TupleValue>`. If we want to implement this feature, we will need to change this to `VecDeque<TupleValue>`. With that change, we will get `push_{back,front}` and `pop_{back,front}` APIs, which we can use to implement this feature. We will also need to check for versionstamp status when we mutate a tuple.

**Comment by ggilley** [comment]
Yes, I think what you are suggesting would work. Let me explain in detail.

If you think of a struct: 
User {
    first_name: String
    last_name: String}
}

So I would have tuples (assuming the value is in the tuple)
    (user, $userid-1, first_name, Sam)
    (user, $userid-1, last_name, Burton)
    (user, $userid-2, first_name, Charles)
    (user, $userid-2, last_name, Barkley)

and I would drop the “user” prefix, then check the next level:

    ($userid-1, first_name, Sam)
    ($userid-1, last_name, Burton)
    ($userid-2, first_name, Charles)
    ($userid-2, last_name, Barkley)

Then find the common prefix, and create the user struct for each:

    for User $userid-1:
        (first_name, Sam)
        (last_name, Burton

   for User $userid-2:
        (first_name, Charles)
        (last_name, Barkley)

So you could think of it as a “slice” operation taking the everything after the first element. pop_front would work for this.

I have a work-around for insert(0) and push() using append().

But your solution would solve both problems.

**Comment by rajivr** [comment]
@ggilley I've pushed a new Tuple API modeled after Rust's VecDeque to a temporary repository [here](https://github.com/rajivr/fdb-tuple-refactor). The new API will land in 0.4.x release and should address your issue. 

I don't know when I might do the next release as am currently busy with the layer crate. Feel free to use this repository in the meantime. Sorry for the delay!
